dscribe/ext/mbtr.cpp

```diff
@@ -642,3 +673,323 @@ vector<map<string, vector<float>>> MBTR::getK3Local(const vector<int> &indices,
     }
     return k3Maps;
 }
+
+void MBTR::getK2Derivatives(py::array_t<float> &derivatives, py::array_t<float> &descriptor, const vector<int> &Z, const vector<vector<float>> &positions, const vector<vector<float>> &distances, const vector<vector<int>> &neighbours, const string &geomFunc, const string &weightFunc, const map<string, float> &parameters, float min, float max, float sigma, int n)
```

Comment:
I completely agree. I now implemented the descriptor and derivative calculations in same functions on both c++ and Python side. Descriptor and derivative arrays are now passed to c++ functions that fill them instead of returning maps that are processed further in Python. In my implementation, c++ operates on dense flattened descriptor arrays, and as a result I changed the `_get_kx`-functions in `mbtr.py` to return dense arrays. The descriptor array is transformed to a sparse array at the end of `create_single`-function if the user requests it.

The change to passing the flattened arrays to c++ comes with a considerable performance boost. The exact performance gains depend on the settings such as the number of grid points and cutoff radii, but I observed computation time dropping by 40-70% when `flatten=True`. When `flatten=False` computation time is unaffected. I observed the same efficiency gains for both `sparse=True` and `sparse=False`.

Reply:
Nice to see some performance comparison. Yes, the idea is to ultimately start using flattened arrays with all descriptors.